music21/interval.py

```diff
@@ -3490,10 +3493,40 @@ def _diatonicTransposePitch(self,
                 # cannot import Accidental here.
                 pitch2.accidental = halfStepsToFix  # type:ignore
 
-            # inherit accidental display type etc. but not current status
-            if pitch2.accidental is not None and pitch1.accidental is not None:
-                pitch2.accidental.inheritDisplay(pitch1.accidental)
-                pitch2.accidental.displayStatus = None  # set accidental display to None
+            # inherit accidental display options
+            accid1: Accidental
+            if pitch1.accidental is None:
+                # Make a real Accidental that represents "non-displayed natural"
+                accid1 = Accidental('natural')
+                accid1.displayStatus = False  # implied by pitch1.accidental is None
+            else:
+                accid1 = pitch1.accidental
+
+            accid2: Accidental
+            if pitch2.accidental is None:
+                # Make a real Accidental that represents "non-displayed natural"
+                accid2 = Accidental('natural')
+                accid2.displayStatus = False  # implied by pitch2.accidental is None
+            else:
+                accid2 = pitch2.accidental
+
+            accid2.inheritDisplay(accid1)
+
+            if pitch2.accidental is None:
+                # accid2 was a made-up Accidental that represents "non-displayed natural",
+                # and then it inherited some display options.  If those display options
+                # still say "non-displayed", we can leave pitch2.accidental as None. But
+                # otherwise we need to set accid2 into pitch2.accidental.
+                if accid2.displayStatus is not False:
+                    pitch2.accidental = accid2
+        else:
+            # no halfStepsToFix, so pitch2 is fine as is, but...
+            # We set pitch2.accidental to None, so we might have lost some
+            # display options. So we restore oldPitch2Accidental if that makes sense.
+            if (pitch2.accidental is None
```

Comment:
pitch2.accidental will always be None here, so this line can be dropped.

But is it true that the displayStatus of the new natural-accidental will always be the same as the old natural-accidental?  That seems like a dangerous assumption.

Reply:
displayStatus inheritance now only happens if `treatAsKeyChange` is True or the Interval is a unison or octave shift.  No-one yet sets `treatAsKeyChange` to True.

Reply:
And now both s.toWrittenPitch and s.toSoundingPitch will end up with treatAsKeyChange True, due to changes in s._transposeByInstrument.  Tests have been modified to test both cases (treatAsKeyChange set and not set).